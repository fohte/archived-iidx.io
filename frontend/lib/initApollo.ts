import 'isomorphic-fetch'

import { InMemoryCache, NormalizedCacheObject } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import { ApolloLink } from 'apollo-link'
import { setContext } from 'apollo-link-context'
import { onError } from 'apollo-link-error'
import { HttpLink } from 'apollo-link-http'

import { auth } from '../lib/firebaseApp'
import isBrowser from './isBrowser'

let apolloClient: ApolloClient<NormalizedCacheObject> | null = null

const handleError = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors) {
    graphQLErrors.map(({ message, locations, path }) =>
      console.error(
        `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
          locations,
        )}, Path: ${path}`,
      ),
    )
  }

  if (networkError) {
    console.error(`[Network error]: ${networkError}`)
  }
})

const withAuthorization = setContext(async (_, { headers }) => {
  if (auth.currentUser) {
    const idToken = await auth.currentUser.getIdToken()

    if (idToken) {
      return {
        headers: { ...headers, authorization: `Bearer ${idToken}` },
      }
    }
  }

  return { headers }
})

const create = (initialState?: NormalizedCacheObject) => {
  const cache = new InMemoryCache().restore(initialState || {})

  const httpLink = new HttpLink({
    uri: process.env.API_URL,
    credentials: 'same-origin',
  })

  return new ApolloClient({
    connectToDevTools: isBrowser,
    ssrMode: !isBrowser,
    link: ApolloLink.from([handleError, withAuthorization, httpLink]),
    cache,
  })
}

const initApollo = (initialState?: NormalizedCacheObject) => {
  // Make sure to create a new client for every server-side request so that data
  // isn't shared between connections (which would be bad)
  if (!isBrowser) {
    return create(initialState)
  }

  // Reuse client on the client-side
  if (!apolloClient) {
    apolloClient = create(initialState)
  }

  return apolloClient
}

export default initApollo
