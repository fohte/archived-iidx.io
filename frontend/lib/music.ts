export const makeTitle = (music: { title: string; subTitle: string }) =>
  `${music.title} ${music.subTitle}`
