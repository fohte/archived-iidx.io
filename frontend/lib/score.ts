import * as _ from 'lodash'

export enum Grade {
  Max = 'MAX',
  A = 'A',
  AA = 'AA',
  AAA = 'AAA',
  B = 'B',
  C = 'C',
  D = 'D',
  E = 'E',
  F = 'F',
}

export const calcScoreRate = (
  score: number,
  numNotes: number,
  precision: number = 2,
) => _.round((score / (numNotes * 2)) * 100, precision)

const getGradeBorders = (numNotes: number): { [key in Grade]: number } => {
  const maxScore = numNotes * 2
  const calcGradeBorder = (coefficients: number): number =>
    Math.ceil(maxScore * coefficients)

  return {
    [Grade.F]: 0,
    [Grade.E]: calcGradeBorder(2 / 9),
    [Grade.D]: calcGradeBorder(3 / 9),
    [Grade.C]: calcGradeBorder(4 / 9),
    [Grade.B]: calcGradeBorder(5 / 9),
    [Grade.A]: calcGradeBorder(6 / 9),
    [Grade.AA]: calcGradeBorder(7 / 9),
    [Grade.AAA]: calcGradeBorder(8 / 9),
    [Grade.Max]: maxScore,
  }
}

export const searchGrade = (
  score: number,
  numNotes: number,
): [Grade | null, number] => {
  if (numNotes * 2 === score) {
    return [Grade.Max, 0]
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([, diff]) => diff >= 0),
    ([, diff]) => Math.abs(diff),
  )

  return nearest || [null, 0]
}

export const searchNextGrade = (
  score: number,
  numNotes: number,
): [Grade | null, number] => {
  if (numNotes * 2 === score) {
    return [Grade.Max, 0]
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([grade, diff]) => diff < 0 && grade !== Grade.F),
    ([, diff]) => Math.abs(diff),
  )

  return nearest || [null, 0]
}
