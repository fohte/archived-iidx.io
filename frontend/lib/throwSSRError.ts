import { ServerResponse } from 'http'

export default (res: ServerResponse | undefined, statusCode: number) => {
  if (res) {
    res.statusCode = statusCode
  }
}
