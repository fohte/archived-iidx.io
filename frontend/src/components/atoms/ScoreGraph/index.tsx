import classnames from 'classnames/bind'
import * as React from 'react'

import { Grade } from '@app/queries'

import * as css from './style.scss'

const cx = classnames.bind(css)

export type Props =
  | {
      loading?: false
      grade: Grade | null
      scoreRate: number
      fullCombo: boolean
    }
  | {
      loading: true
    }

const ScoreGraph: React.SFC<Props> = props => {
  if (props.loading) {
    return <div className={cx('wrapper', 'loading')} />
  } else {
    const { grade, scoreRate, fullCombo } = props
    return (
      <div className={cx('wrapper')}>
        <div
          className={cx('progress', {
            'grade-aaa': grade === Grade.Aaa,
            'grade-aa': grade === Grade.Aa,
            'grade-a': grade === Grade.A,
            'grade-b': grade === Grade.B,
            'grade-c': grade === Grade.C,
            'grade-d': grade === Grade.D,
            'grade-e': grade === Grade.E,
            'grade-f': grade === Grade.F,
            'full-combo': fullCombo,
          })}
          style={{ width: `${scoreRate * 100}%` }}
        />
        <div className={cx('target', 'grade-a')} />
        <div className={cx('target', 'grade-aa')} />
        <div className={cx('target', 'grade-aaa')} />
      </div>
    )
  }
}

export default ScoreGraph
