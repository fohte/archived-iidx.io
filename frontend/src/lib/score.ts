import * as _ from 'lodash'

export enum Grade {
  Max = 'MAX',
  A = 'A',
  AA = 'AA',
  AAA = 'AAA',
  B = 'B',
  C = 'C',
  D = 'D',
  E = 'E',
  F = 'F',
}

export interface GradeDiff {
  grade: Grade | null
  diff: number
}

export const calcScoreRate = (
  score: number,
  numNotes: number,
  precision: number = 2,
) => _.round((score / (numNotes * 2)) * 100, precision)

export const calcGradeBorder = (coefficients: number, numNotes: number) =>
  Math.ceil(numNotes * 2 * coefficients)

export const getGradeBorders = (
  numNotes: number,
): { [key in Grade]: number } => ({
  [Grade.F]: 0,
  [Grade.E]: calcGradeBorder(numNotes, 2 / 9),
  [Grade.D]: calcGradeBorder(numNotes, 3 / 9),
  [Grade.C]: calcGradeBorder(numNotes, 4 / 9),
  [Grade.B]: calcGradeBorder(numNotes, 5 / 9),
  [Grade.A]: calcGradeBorder(numNotes, 6 / 9),
  [Grade.AA]: calcGradeBorder(numNotes, 7 / 9),
  [Grade.AAA]: calcGradeBorder(numNotes, 8 / 9),
  [Grade.Max]: numNotes * 2,
})

export const defaultGradeDiff: GradeDiff = {
  grade: null,
  diff: 0,
}

export const searchGrade = (score: number, numNotes: number): GradeDiff => {
  if (numNotes * 2 === score) {
    return { grade: Grade.Max, diff: 0 }
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([, diff]) => diff >= 0),
    ([, diff]) => Math.abs(diff),
  )

  if (!nearest) {
    return defaultGradeDiff
  }

  return { grade: nearest[0], diff: nearest[1] }
}

export const searchNextGrade = (score: number, numNotes: number): GradeDiff => {
  if (numNotes * 2 === score) {
    return { grade: Grade.Max, diff: 0 }
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([grade, diff]) => diff < 0 && grade !== Grade.F),
    ([, diff]) => Math.abs(diff),
  )

  if (!nearest) {
    return defaultGradeDiff
  }

  return { grade: nearest[0], diff: nearest[1] }
}

export const isMax = ({ grade, diff }: GradeDiff): boolean =>
  grade === Grade.Max && diff === 0
