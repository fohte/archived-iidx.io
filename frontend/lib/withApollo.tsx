import { NormalizedCacheObject } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import Head from 'next/head'
import React from 'react'
import { ApolloProvider, getDataFromTree } from 'react-apollo'
import { setDisplayName, wrapDisplayName } from 'recompose'

import initApollo from './initApollo'
import isBrowser from './isBrowser'

interface Props {
  serverState: {
    apollo: {
      data: NormalizedCacheObject
    }
  }
}

export default ComposedComponent => {
  class WithData extends React.Component<Props> {
    public static async getInitialProps(ctx) {
      // Initial serverState with apollo (empty)
      let serverState: Props['serverState'] = {
        apollo: {
          data: {},
        },
      }

      // Evaluate the composed component's getInitialProps()
      let composedInitialProps = {}
      if (ComposedComponent.getInitialProps) {
        composedInitialProps = await ComposedComponent.getInitialProps(ctx)
      }

      // Run all GraphQL queries in the component tree
      // and extract the resulting data
      if (!isBrowser) {
        const apollo = initApollo()

        try {
          // Run all GraphQL queries
          await getDataFromTree(
            <ApolloProvider client={apollo}>
              <ComposedComponent {...composedInitialProps} />
            </ApolloProvider>,
            {
              router: {
                asPath: ctx.asPath,
                pathname: ctx.pathname,
                query: ctx.query,
              },
            },
          )
        } catch (error) {
          // Prevent Apollo Client GraphQL errors from crashing SSR.
          // Handle them in components via the data.error prop:
          // https://www.apollographql.com/docs/react/api/react-apollo.html#graphql-query-data-error
        }
        // getDataFromTree does not call componentWillUnmount
        // head side effect therefore need to be cleared manually
        Head.rewind()

        // Extract query data from the Apollo store
        serverState = {
          apollo: {
            data: apollo.cache.extract(),
          },
        }
      }

      return {
        serverState,
        ...composedInitialProps,
      }
    }

    public apollo: ApolloClient<NormalizedCacheObject>

    constructor(props) {
      super(props)
      this.apollo = initApollo(this.props.serverState.apollo.data)
    }

    public render() {
      return (
        <ApolloProvider client={this.apollo}>
          <ComposedComponent {...this.props} />
        </ApolloProvider>
      )
    }
  }

  return setDisplayName(wrapDisplayName(WithData, 'WithData'))(WithData)
}
