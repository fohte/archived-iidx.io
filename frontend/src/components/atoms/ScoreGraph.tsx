import * as React from 'react'
import styled from 'styled-components'
import { Diff } from 'utility-types'

import { Grade } from '@app/lib/score'
import { colors } from '@app/styles'

export interface Props {
  grade: Grade | null
  scoreRate: number
}

const Wrapper = styled.div`
  position: relative;
  width: 100%;
  height: 1.5em;
`

interface ScoreBarProps {
  rate: number
  color: string
}

const ScoreBar = styled('div')<ScoreBarProps>`
  width: ${props => props.rate}%;
  height: 50%;
  background-color: ${props => props.color};
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
`

interface VerticalBarProps {
  length: number
  borderStyle: string
  borderColor: string
}

const VerticalBar = styled('div')<VerticalBarProps>`
  width: ${props => props.length}%;
  height: 100%;
  border-right: 1px ${props => props.borderStyle} ${props => props.borderColor};
  position: absolute;
`

interface TargetBarProps {
  isReached: boolean
  length: number
}

const TargetBar = styled(VerticalBar).attrs<
  TargetBarProps,
  Diff<VerticalBarProps, TargetBarProps>
>(({ isReached }) => ({
  borderColor: isReached ? colors.base.blue : colors.base.grey,
  borderStyle: 'dotted',
}))<TargetBarProps>`
  z-index: 10;
`

interface BorderBarProps {}

const BorderBar = styled(VerticalBar).attrs<
  BorderBarProps,
  Diff<VerticalBarProps, BorderBarProps>
>(() => ({
  borderColor: colors.base.grey,
  borderStyle: 'solid',
  length: 100,
}))`
  z-index: 10;
  border-left: 1px ${props => props.borderStyle} ${props => props.borderColor};
`

const ScoreGraph: React.SFC<Props> = ({ grade, scoreRate }) => (
  <Wrapper>
    <ScoreBar rate={scoreRate} color={colors.grade[grade || 'default']} />
    <BorderBar />
    {[6, 7, 8].map(x => {
      const targetRate = (x / 9) * 100
      return (
        <TargetBar
          key={x}
          length={targetRate}
          isReached={scoreRate >= targetRate}
        />
      )
    })}
  </Wrapper>
)

export default ScoreGraph
