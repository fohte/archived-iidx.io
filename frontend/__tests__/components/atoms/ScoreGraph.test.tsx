import * as React from 'react'
import * as renderer from 'react-test-renderer'

import ScoreGraph, { Props } from '@app/components/atoms/ScoreGraph'
import { Grade } from '@app/lib/score'

describe('ScoreGraph', () => {
  const params: Props[] = [
    { grade: Grade.Max, scoreRate: 100 },
    { grade: Grade.AAA, scoreRate: Math.ceil((8 / 9) * 100) },
    { grade: Grade.AA, scoreRate: Math.ceil((7 / 9) * 100) },
    { grade: Grade.A, scoreRate: Math.ceil((6 / 9) * 100) },
    { grade: Grade.B, scoreRate: Math.ceil((5 / 9) * 100) },
    { grade: Grade.C, scoreRate: Math.ceil((4 / 9) * 100) },
    { grade: Grade.D, scoreRate: Math.ceil((3 / 9) * 100) },
    { grade: Grade.E, scoreRate: Math.ceil((2 / 9) * 100) },
    { grade: Grade.F, scoreRate: Math.ceil((1 / 9) * 100) },
    { grade: null, scoreRate: 0 },
  ]

  params.forEach(({ grade, scoreRate }) => {
    it(`renders correctly when grade is ${grade} and score rate is ${scoreRate}`, () => {
      const tree = renderer
        .create(<ScoreGraph grade={grade} scoreRate={scoreRate} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
