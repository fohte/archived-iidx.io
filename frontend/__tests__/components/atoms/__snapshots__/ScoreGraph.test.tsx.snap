// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreGraph renders correctly when grade is A and score rate is 67 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 67%;
  height: 50%;
  background-color: #2185D0;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#2185D0"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is AA and score rate is 78 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 78%;
  height: 50%;
  background-color: #B5CC18;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#B5CC18"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is AAA and score rate is 89 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 89%;
  height: 50%;
  background-color: #F2711C;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#F2711C"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is B and score rate is 56 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 56%;
  height: 50%;
  background-color: #1B1C1D;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#1B1C1D"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is C and score rate is 45 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 45%;
  height: 50%;
  background-color: #1B1C1D;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#1B1C1D"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is D and score rate is 34 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 34%;
  height: 50%;
  background-color: #1B1C1D;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#1B1C1D"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is E and score rate is 23 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 23%;
  height: 50%;
  background-color: #1B1C1D;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#1B1C1D"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is F and score rate is 12 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 12%;
  height: 50%;
  background-color: #1B1C1D;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#1B1C1D"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is MAX and score rate is 100 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 100%;
  height: 50%;
  background-color: #DB2828;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #2185D0;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#DB2828"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;

exports[`ScoreGraph renders correctly when grade is null and score rate is 0 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: 1.5em;
}

.c1 {
  width: 0%;
  height: 50%;
  background-color: #1B1C1D;
  position: absolute;
  vertical-align: middle;
  top: 25%;
  z-index: 0;
}

.c5 {
  width: 88.88888888888889%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c4 {
  width: 77.77777777777779%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c3 {
  width: 66.66666666666666%;
  height: 100%;
  border-right: 1px dotted #767676;
  position: absolute;
  z-index: 10;
}

.c2 {
  width: 100%;
  height: 100%;
  border-right: 1px solid #767676;
  position: absolute;
  z-index: 10;
  border-left: 1px solid #767676;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#1B1C1D"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
</div>
`;
